animation_nodes/nodes/spline/splines_from_edges_utils.pyx

```diff
@@ -0,0 +1,117 @@
+    Mesh, Vector3DList, EdgeIndicesList, PolySpline,
+    VirtualDoubleList, FloatList
+)
+
+def splinesFromBranches(Vector3DList vertices, EdgeIndicesList edges, VirtualDoubleList radii):
+    cdef int i, j
+    cdef int edgesAmount = len(edges)
+    cdef int verticesAmount = len(vertices)
+
+    # Compute how many neighbour each vertex have.
+    cdef int *neighboursAmounts = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    memset(neighboursAmounts, 0, verticesAmount * sizeof(int))
+    for i in range(edgesAmount):
+        neighboursAmounts[edges.data[i].v1] += 1
+        neighboursAmounts[edges.data[i].v2] += 1
+
+    # Compute the start index of each group of neighbours of each vertex.
+    cdef int *neighboursStarts = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    cdef int start = 0
+    for i in range(verticesAmount):
+        neighboursStarts[i] = start
+        start += neighboursAmounts[i]
+
+    # Keep track of how many index is in each group of neighbours at each iteration.
+    cdef int *filledSpaces = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    memset(filledSpaces, 0, verticesAmount * sizeof(int))
+
+    # Compute the indices of neighbouring vertices of each vertex.
+    cdef int *neighbours = <int*>PyMem_Malloc(edgesAmount * 2 * sizeof(int))
+    for i in range(edgesAmount):
+        neighbours[neighboursStarts[edges.data[i].v1] + filledSpaces[edges.data[i].v1]] = edges.data[i].v2
+        filledSpaces[edges.data[i].v1] += 1
+
+        neighbours[neighboursStarts[edges.data[i].v2] + filledSpaces[edges.data[i].v2]] = edges.data[i].v1
+        filledSpaces[edges.data[i].v2] += 1
+
+    # Find the indices of the vertices that are not connected to two vertices.
+    cdef int *nonBipolarVertices = <int*>PyMem_Malloc(verticesAmount * sizeof(int))
+    cdef int nonBipolarVertsCount = 0
+    for i in range(verticesAmount):
+        if neighboursAmounts[i] != 2:
+            nonBipolarVertices[nonBipolarVertsCount] = i
+            nonBipolarVertsCount += 1
+    nonBipolarVertices = <int*>PyMem_Realloc(nonBipolarVertices, nonBipolarVertsCount * sizeof(int))
+
+    # Generate Splines.
+    cdef list splines = []
+    cdef PolySpline spline
+    cdef FloatList splineRadii
+    cdef Vector3DList splineVertices
+    cdef int nonBipolarVertex, currentVertex, nextVertex
+    for i in range(nonBipolarVertsCount):
+        nonBipolarVertex = nonBipolarVertices[i]
+        for j in range(neighboursAmounts[nonBipolarVertex]):
+            nextVertex = neighbours[neighboursStarts[nonBipolarVertex] + j]
+            if (filledSpaces[nextVertex] == neighboursAmounts[nextVertex] or
+                                           neighboursAmounts[nextVertex] != 2):
+                splineVertices = Vector3DList.__new__(Vector3DList, capacity = verticesAmount)
+                splineRadii = FloatList.__new__(FloatList, capacity = verticesAmount)
+
+                splineVertices.append(vertices[nonBipolarVertex])
```

Comment:
using the `append_LowLevel` function here might result in a noticable performance improvement

Reply:
I am wondering if it would be better to do:
```python
splineRadii = FloatList.__new__(FloatList, length = verticesAmount, capacity = verticesAmount)
splineRadii[0] = radii.get(nonBipolarVertex)
splineRadii[1] = radii.get(nextVertex)
count = 2
for _ in range():
    ...
    splineRadii[count] = radii.get(nextVertex)
    count += 1
```

Reply:
that would be even faster, but not sure if it is worth the extra variable

btw, when you specify the length of a list in the constructor, you don't need to specify the capacity (if it is the same)

Reply:
We could reuse the variable `j`, then we won't have to create any extra variables. And it will be used in a narrow vertical scope. Should we do that?

Reply:
I meant to use it instead of `count`, I was responding to your comment:
> but not sure if it is worth the extra variable.
But yah, I will no longer reuse variables, will have to edit some stuff in account with this. :D